### PR TITLE
Rewrites: in_array() versus isset()

### PR DESCRIPTION
**Pull request by Quix0r**
As mentioned in #81 in_array() performs slow as it searches the whole array for a value. If you can use isset() because the value is unimportant, only the fact that `$foo` is in `$foos`, then you should use `isset()`. If you need to check on NULL values, use `array_key_exists()` (which we don't need/do here).

So isset() is IMHO really fine and better here.

### CODE REVIEW

File: src/PhpSpreadsheet/CachedObjectStorageFactory.php [hunk 1]
```diff
@@ -136,7 +136,7 @@ public static function getCacheStorageClass()
      **/
     public static function getAllCacheStorageMethods()
     {
-        return self::$storageMethods;
+        return array_keys(self::$storageMethods);
```
Comment (Quix0r) [inline src/PhpSpreadsheet/CachedObjectStorageFactory.php]: This returns the same array as expected (no change of "public method behaviour").

File: src/PhpSpreadsheet/Writer/Xls/Worksheet.php [hunk 2]
```diff
@@ -518,8 +518,10 @@ public function close()
                 foreach ($conditionalStyles as $conditional) {
                     if ($conditional->getConditionType() == \PhpOffice\PhpSpreadsheet\Style\Conditional::CONDITION_EXPRESSION
                         || $conditional->getConditionType() == \PhpOffice\PhpSpreadsheet\Style\Conditional::CONDITION_CELLIS) {
-                        if (!in_array($conditional->getHashCode(), $arrConditional)) {
-                            $arrConditional[] = $conditional->getHashCode();
+                        if (!isset($arrConditional[$conditional->getHashCode()])) {
+                            // This hash code has been handled
+                            $arrConditional[$conditional->getHashCode()] = true;
```
Comment (Quix0r) [inline src/PhpSpreadsheet/Writer/Xls/Worksheet.php]: This (old) code should make sure that $condinal is only written once, as I understand the code.

**Comment by Quix0r** [comment]
Hmm a lot merges against upstream/develop, maybe I should rebase (which will close this PR)?

**Comment by PowerKiKi** [comment]
Yes you should rebase. And it will not close the PR, but you may have to force push (which is fine for this).

**Comment by Quix0r** [comment]
Waahh, tons of merge conflicts to resolve ... Any idea how to make it faster?

`git rebase -m upstream/develop`

I use this commonly.

**Comment by PowerKiKi** [comment]
Since your patch is not that big, it may be faster to start from scratch.

But more importantly, did you benchmark those changes? At least the cache part seems to make no significant difference while making the code more complex to follow. I'll have to double check once I'm on my laptop, but I'm not sure I would merge it as is, unless you can show a significant speed improvement for normal-ish use-cases.

**Comment by Quix0r** [comment]
Redone and force-pushed ...

**Comment by PowerKiKi** [comment]
> No need to benchmark

In the case of `CachedObjectStorageFactory` you tried to increase speed of `initialize()`, at the cost of code readability of overall class, and a possible decrease in speed for `getAllCacheStorageMethods()` and `getCacheStorageMethods()`. Both of those methods are public and it is hard to predict how they are used in the wild.

So it seems you tried to improve 1 thing by possibly worsening 3 things, without a strong argument in favor of this change. To me it doesn't look like an improvement. Especially because the name `initialize()` suggest that it should probably not be used very often, hence does not have a significative impact in the first place.

For those reasons, I only merged the two other classes.

**Comment by Quix0r** [comment]
I may come back with more changes where `isset()` can be safely used.

**Pull request closed without merging**